### PR TITLE
Create rustdoc_internals feature gate

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
As suggested by @camelid [here](https://github.com/rust-lang/rust/pull/90398#issuecomment-955093851), since `doc_keyword` and `doc_primitive` aren't meant to be stabilized, we could put them behind a same feature flag.

This is pretty much what it would look like (needs to update the tests too).

The tracking issue is https://github.com/rust-lang/rust/issues/90418.

What do you think @rust-lang/rustdoc ?

### CODE REVIEW

File: compiler/rustc_feature/src/active.rs [hunk 1]
```diff
@@ -688,6 +690,10 @@ declare_features! (
     /// not changed from prior instances of the same struct (RFC #2528)
     (incomplete, type_changing_struct_update, "1.58.0", Some(86555), None),
 
+    /// Allows using internal rustdoc features like doc(primitive) or doc(keyword)
+    /// without a future-incompat warning.
```
Comment (camelid) [inline compiler/rustc_feature/src/active.rs:694]: `doc(keyword)` on stable is a hard error, only `doc(primitive)` is a warning. Also, please use inline code ;)

File: compiler/rustc_ast_passes/src/feature_gate.rs [hunk 2]
```diff
@@ -322,7 +322,8 @@ impl<'a> Visitor<'a> for PostExpansionVisitor<'a> {
                     cfg_hide => doc_cfg_hide
                     masked => doc_masked
                     notable_trait => doc_notable_trait
-                    keyword => doc_keyword
+                    keyword => rustdoc_internals
+                    primitive => rustdoc_internals
```
Comment (camelid) [inline compiler/rustc_ast_passes/src/feature_gate.rs:326]: I'm pretty sure this will make `doc(primitive)` a hard error. Where is feature-gating of `doc(primitive)` currently handled?

Comment (jyn514) [reply compiler/rustc_ast_passes/src/feature_gate.rs:326]: It's in `check_attr.rs`.
```
[nix-shell:~/rustc3]$ rg doc_primitive compiler/
compiler/rustc_feature/src/active.rs
664:    (active, doc_primitive, "1.56.0", Some(88070), None),

compiler/rustc_passes/src/check_attr.rs
965:                            if !self.tcx.features().doc_primitive {
```

**Comment by jyn514** [comment]
This seems ... kind of unnecessary to me? We can document on the tracking issues that these aren't meant to be stabilized, I don't think it also needs to be in the source code of everything using them. And as you've noticed the logic for `doc_primitive` is pretty different from that of `doc_keyword`, so I don't think it's particularly useful to combine them.

**Comment by camelid** [comment]
> This seems ... kind of unnecessary to me? We can document on the tracking issues that these aren't meant to be stabilized, I don't think it also needs to be in the source code of everything using them.

The standard library often uses the naming convention of `X_internals` to indicate that something is perma-unstable. Whereas `#![feature(doc_keyword)]` doesn't give any indication that it's only meant for internal use. People don't necessarily look at tracking issues, so IMO it's best to indicate in the feature name that something is internal-only. Another example is `#[rustc_layout(debug)]` which is under the feature-gate `rustc_attrs`, not something like `debug_layout`.

If merging the feature gates would be too complicated, I'm okay to leave them separate, but I would prefer to merge them and make it clear in the name that they are internal-only.

> I don't think it also needs to be in the source code of everything using them.

They're only used in the standard library and rustdoc IIUC, and this would reduce the number of feature gates necessary?

**Comment by jyn514** [comment]
Hmm, ok, seems reasonable.

File: src/test/ui/feature-gates/feature-gate-rustdoc_internals.stderr [hunk 3]
```diff
@@ -0,0 +1,12 @@
+error[E0658]: `#[doc(keyword)]` is experimental
```
Comment (camelid) [inline src/test/ui/feature-gates/feature-gate-rustdoc_internals.stderr:1]: Can you also change this message to something like the following?

```suggestion
error[E0658]: `#[doc(keyword)]` is meant for internal use only
```

**Comment by GuillaumeGomez** [comment]
Updated the message and set the PR as ready to review.

File: compiler/rustc_feature/src/active.rs [hunk 4]
```diff
@@ -417,6 +417,7 @@ declare_features! (
     /// Allows `'a: { break 'a; }`.
     (active, label_break_value, "1.28.0", Some(48594), None),
 
+    // FIXME: to be removed once rustdoc_internals is merged.
```
Comment (camelid) [inline compiler/rustc_feature/src/active.rs:420]: (This and the below FIXME still need to be fixed.)

Comment (GuillaumeGomez) [reply compiler/rustc_feature/src/active.rs:420]: We need this PR to be merged first to do so (and wait for a new nightly). :)

Comment (camelid) [reply compiler/rustc_feature/src/active.rs:420]: Why do we need to wait?

Comment (GuillaumeGomez) [reply compiler/rustc_feature/src/active.rs:420]: Because the stage 0 is compiled with an old rustc version which doesn't have that feature yet. So we need to keep the old ones around until the nightly used for stage 0 is updated.

Comment (camelid) [reply compiler/rustc_feature/src/active.rs:420]: I don't think that's necessary. You can just do this:

```rust
#![cfg_attr(bootstrap, feature(doc_keyword)]
#![cfg_attr(not(bootstrap), feature(rustdoc_internals)]
```

(You may have to swap the `bootstrap` and `not(bootstrap)`; I always get them mixed up.)

Comment (GuillaumeGomez) [reply compiler/rustc_feature/src/active.rs:420]: I tried and it failed. Also, the problem is mainly in `libcore` and `libstd` which (unless I'm wrong) needs an older rustc in any case. At least I tried with `bootstrap` and failed. But maybe I simply inverted the bootstrap?

**Comment by GuillaumeGomez** [comment]
Updated!

**Comment by GuillaumeGomez** [comment]
Updated!

**Comment by GuillaumeGomez** [comment]
@camelid I think my comment wasn't added so: you were right, I inverted the `cfg_attr` condition for `bootstrap`. Works much better now. :)

File: library/core/src/lib.rs [hunk 5]
```diff
@@ -161,7 +161,10 @@
 #![feature(decl_macro)]
 #![feature(doc_cfg)]
 #![feature(doc_notable_trait)]
-#![feature(doc_primitive)]
+// FIXME: to be removed once rustdoc_internals is merged.
```
Comment (camelid) [inline library/core/src/lib.rs:164]: Please remove these FIXMEs. Whoever updates the bootstrap compiler will update all the uses of `cfg(bootstrap)` :)

File: src/test/ui/feature-gates/feature-gate-rustdoc_internals.rs [hunk 6]
```diff
@@ -0,0 +1,5 @@
+#[doc(keyword = "match")] //~ ERROR: `#[doc(keyword)]` is meant for internal use only
```
Comment (camelid) [inline src/test/ui/feature-gates/feature-gate-rustdoc_internals.rs:1]: Why did you add this test? It looks identical to `feature-gate-doc_keyword.rs`. I think it'd be better to just rename the `doc_keyword` test to `rustdoc_internals`.

File: src/test/ui/feature-gates/feature-gate-rustdoc_internals.stderr [hunk 7]
```diff
@@ -0,0 +1,12 @@
+error[E0658]: `#[doc(keyword)]` is meant for internal use only
```
Comment (camelid) [inline src/test/ui/feature-gates/feature-gate-rustdoc_internals.stderr:1]: Can you add a test to make sure the "feature gate renamed error" is shown for `#![feature(doc_keyword)]` and `doc_primitive`?

File: compiler/rustc_ast_passes/src/feature_gate.rs [hunk 8]
```diff
@@ -325,8 +325,12 @@ impl<'a> Visitor<'a> for PostExpansionVisitor<'a> {
                     cfg_hide => doc_cfg_hide
                     masked => doc_masked
                     notable_trait => doc_notable_trait
-                    keyword => doc_keyword
                 );
+
+                if nested_meta.has_name(sym::keyword) {
+                    let msg = concat!("`#[doc(keyword)]` is meant for internal use only");
```
Comment (camelid) [inline compiler/rustc_ast_passes/src/feature_gate.rs:331]: Hmm, maybe just allow a custom message to be chosen with the macro? I worry this will fall out of sync with the macro.

File: src/test/ui/feature-gates/feature-gate-rustdoc_internals.rs [hunk 9]
```diff
@@ -0,0 +1,5 @@
...
+/// wonderful
+mod foo{}
```
Comment (camelid) [inline src/test/ui/feature-gates/feature-gate-rustdoc_internals.rs:3]: nit (feel free to ignore):

```suggestion
mod foo {}
```

File: src/test/ui/rustdoc/renamed_features.rs [hunk 10]
```diff
@@ -0,0 +1,5 @@
+#![feature(doc_keyword)] //~ ERROR
```
Comment (camelid) [inline src/test/ui/rustdoc/renamed_features.rs:1]: Could you rename this test to something a bit more descriptive, that includes rustdoc_internals in it?

**Comment by GuillaumeGomez** [comment]
> Could you rename this test to something a bit more descriptive, that includes rustdoc_internals in it?

Any suggestion? I'm very bad at naming...

**Comment by camelid** [comment]
> > Could you rename this test to something a bit more descriptive, that includes rustdoc_internals in it?
> 
> Any suggestion? I'm very bad at naming...

I think this should be good enough: `renamed-features-rustdoc_internals.rs`

**Comment by GuillaumeGomez** [comment]
Ok!

**Comment by GuillaumeGomez** [comment]
Updated the name.

File: compiler/rustc_feature/src/active.rs [hunk 11]
```diff
@@ -161,6 +161,9 @@ declare_features! (
     (active, profiler_runtime, "1.18.0", None, None),
     /// Allows using `rustc_*` attributes (RFC 572).
     (active, rustc_attrs, "1.0.0", None, None),
+    /// Allows using internal rustdoc features like `doc(primitive)` or `doc(keyword)`
+    /// without a future-incompat warning.
```
Comment (camelid) [inline compiler/rustc_feature/src/active.rs:165]: Only `doc(primitive)` is a future-incompat warning. To avoid this getting out of date, let's just do this:

```suggestion
    /// Allows using internal rustdoc features like `doc(primitive)` or `doc(keyword)`.
```

**Comment by GuillaumeGomez** [comment]
Done!

File: compiler/rustc_feature/src/active.rs [hunk 12]
```diff
@@ -161,6 +161,8 @@ declare_features! (
     (active, profiler_runtime, "1.18.0", None, None),
     /// Allows using `rustc_*` attributes (RFC 572).
     (active, rustc_attrs, "1.0.0", None, None),
+    /// Allows using internal rustdoc features like `doc(primitive)` or `doc(keyword)`.
+    (active, rustdoc_internals, "1.58.0", Some(90418), None),
```
Comment (camelid) [inline compiler/rustc_feature/src/active.rs:165]: This is in the wrong section; it should be next to doc_notable_trait.

Comment (camelid) [reply compiler/rustc_feature/src/active.rs:165]: Also, can you create an actual tracking issue? We shouldn't just link to this PR.

File: compiler/rustc_feature/src/removed.rs [hunk 13]
```diff
@@ -76,6 +76,12 @@ declare_features! (
     /// Allows the use of `#[derive(Anything)]` as sugar for `#[derive_Anything]`.
     (removed, custom_derive, "1.32.0", Some(29644), None,
      Some("subsumed by `#[proc_macro_derive]`")),
+    /// Allows using `#[doc(keyword = "...")]`.
+    (removed, doc_keyword, "1.28.0", Some(51315), None,
+     Some("merged into `#![feature(rustdoc_internals)]`")),
+    /// Allows using doc(primitive) without a future-incompat warning
```
Comment (camelid) [inline compiler/rustc_feature/src/removed.rs:82]: nit:

```suggestion
    /// Allows using `doc(primitive)` without a future-incompat warning
```

**Review by camelid — APPROVED** [review]
Left a few comments. r=me after fixing those.

**Comment by GuillaumeGomez** [comment]
> Also, can you create an actual tracking issue? We shouldn't just link to this PR.

It does link to the tracking issue. Unless I missed something?

Otherwise, I updated your comments.

**Comment by camelid** [comment]
Sorry, maybe GitHub glitched and showed me the wrong page. Tracking issue looks good to me.

File: compiler/rustc_feature/src/active.rs [hunk 14]
```diff
@@ -206,6 +206,8 @@ declare_features! (
     (active, rustc_allow_const_fn_unstable, "1.49.0", Some(69399), None),
     /// Allows using compiler's own crates.
     (active, rustc_private, "1.0.0", Some(27812), None),
+        /// Allows using internal rustdoc features like `doc(primitive)` or `doc(keyword)`.
```
Comment (camelid) [inline compiler/rustc_feature/src/active.rs:209]:
```suggestion
    /// Allows using internal rustdoc features like `doc(primitive)` or `doc(keyword)`.
```

File: compiler/rustc_feature/src/removed.rs [hunk 15]
```diff
@@ -76,6 +76,12 @@ declare_features! (
     /// Allows the use of `#[derive(Anything)]` as sugar for `#[derive_Anything]`.
     (removed, custom_derive, "1.32.0", Some(29644), None,
      Some("subsumed by `#[proc_macro_derive]`")),
+    /// Allows using `#[doc(keyword = "...")]`.
+    (removed, doc_keyword, "1.28.0", Some(51315), None,
+     Some("merged into `#![feature(rustdoc_internals)]`")),
+    /// Allows using `doc(primitive)` without a future-incompat warning
```
Comment (camelid) [inline compiler/rustc_feature/src/removed.rs:82]:
```suggestion
    /// Allows using `doc(primitive)` without a future-incompat warning.
```

**Review by camelid — APPROVED** [review]
r=me with these two comments fixed

File: compiler/rustc_feature/src/removed.rs (hunk 15 above)
Comment (camelid) [reply compiler/rustc_feature/src/removed.rs:82]: (Totally optional to fix this by the way.)

**Comment by GuillaumeGomez** [comment]
@camelid Your comments made perfect sense (and I have no clue how I messed up the indent...). Updated!

**Comment by camelid** [comment]
Thanks!

@bors r+ rollup
